### PR TITLE
fix flake: propagation with taint and toleration testing

### LINKED ISSUE #614: flake: propagation with taint and toleration testing
#### Which jobs are flaking:
E2E
https://github.com/karmada-io/karmada/runs/3346097576

#### Which test(s) are flaking:
[Fail] propagation with taint and toleration testing Deployment propagation testing [BeforeEach] deployment with cluster tolerations testing 
/root/actions-runner/_work/karmada/karmada/test/e2e/tainttoleration_test.go:81

#### Reason for failure:
```
propagation with taint and toleration testing Deployment propagation testing 
  deployment with cluster tolerations testing
  /root/actions-runner/_work/karmada/karmada/test/e2e/tainttoleration_test.go:107
STEP: creating policy(karmadatest-r24/deploy-tph)
STEP: adding taints to clusters
I0817 09:48:54.769198 3547034 tainttoleration_test.go:78] Adding taints(cluster-toleration.karmada.io=member1:NoSchedule) to cluster(member1)
I0817 09:48:54.800076 3547034 tainttoleration_test.go:78] Adding taints(cluster-toleration.karmada.io=member2:NoSchedule) to cluster(member2)
STEP: removing policy(karmadatest-r24/deploy-tph)
STEP: removing taints in cluster

• Failure in Spec Setup (BeforeEach) [0.150 seconds]
propagation with taint and toleration testing
/root/actions-runner/_work/karmada/karmada/test/e2e/tainttoleration_test.go:20
  Deployment propagation testing
  /root/actions-runner/_work/karmada/karmada/test/e2e/tainttoleration_test.go:21
    deployment with cluster tolerations testing [BeforeEach]
    /root/actions-runner/_work/karmada/karmada/test/e2e/tainttoleration_test.go:107

    Unexpected error:
        <*errors.StatusError | 0xc0005b8960>: {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "Failure",
                Message: "Operation cannot be fulfilled on clusters.cluster.karmada.io \"member2\": the object has been modified; please apply your changes to the latest version and try again",
                Reason: "Conflict",
                Details: {
                    Name: "member2",
                    Group: "cluster.karmada.io",
                    Kind: "clusters",
                    UID: "",
                    Causes: nil,
                    RetryAfterSeconds: 0,
                },
                Code: 409,
            },
        }
        Operation cannot be fulfilled on clusters.cluster.karmada.io "member2": the object has been modified; please apply your changes to the latest version and try again
    occurred

    /root/actions-runner/_work/karmada/karmada/test/e2e/tainttoleration_test.go:81
```

#### Anything else we need to know:
We should update the cluster with `retry` to solve the possible conflics.

/kind flake

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
Signed-off-by: changzhen <changzhen5@huawei.com>

**What type of PR is this?**

/kind cleanup
/kind flake

**What this PR does / why we need it**:

Update the cluster with retry to solve the possible conflics in #614 .

```
------------------------------
propagation with taint and toleration testing Deployment propagation testing 
  deployment with cluster tolerations testing
  /root/cz/karmada/test/e2e/tainttoleration_test.go:131
STEP: creating policy(karmadatest-29n/deploy-p5z)
STEP: adding taints to clusters
I0820 15:54:27.023959  224122 tainttoleration_test.go:72] Adding taints(cluster-toleration.karmada.io=member1:NoSchedule) to clusterName(member1)
I0820 15:54:27.053125  224122 tainttoleration_test.go:72] Adding taints(cluster-toleration.karmada.io=member2:NoSchedule) to clusterName(member2)
I0820 15:54:27.071649  224122 tainttoleration_test.go:72] Adding taints(cluster-toleration.karmada.io=member3:NoSchedule) to clusterName(member3)
STEP: creating deployment(karmadatest-29n/deploy-p5z)
STEP: check if deployment(karmadatest-29n/deploy-p5z) only scheduled to tolerated cluster(member1)
I0820 15:54:32.180906  224122 failover_test.go:236] The ResourceBinding(karmadatest-29n/deploy-p5z-deployment) schedule result is: member1
STEP: removing deployment(karmadatest-29n/deploy-p5z)
STEP: removing policy(karmadatest-29n/deploy-p5z)
STEP: removing taints in cluster
I0820 15:54:32.198348  224122 tainttoleration_test.go:117] update taints([]) to clusterName(member1)
I0820 15:54:32.220473  224122 tainttoleration_test.go:117] update taints([]) to clusterName(member2)
I0820 15:54:32.248542  224122 tainttoleration_test.go:117] update taints([]) to clusterName(member3)

• [SLOW TEST:5.245 seconds]
propagation with taint and toleration testing
/root/cz/karmada/test/e2e/tainttoleration_test.go:21
  Deployment propagation testing
  /root/cz/karmada/test/e2e/tainttoleration_test.go:22
    deployment with cluster tolerations testing
    /root/cz/karmada/test/e2e/tainttoleration_test.go:131
------------------------------
```
**Which issue(s) this PR fixes**:
Fixes #614

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/tainttoleration_test.go [hunk 1]
```diff
@@ -59,26 +60,34 @@ var _ = ginkgo.Describe("propagation with taint and toleration testing", func()
...
-							Value:  clusterObj.Name,
+							Value:  clusterName,
 							Effect: corev1.TaintEffectNoSchedule,
 						},
 					}
-
-					for _, taint := range clusterObj.Spec.Taints {
-						klog.Infof("Adding taints(%s) to cluster(%s)", taint.ToString(), clusterObj.Name)
+					for _, taint := range taints {
+						klog.Infof("Adding taints(%s) to clusterName(%s)", taint.ToString(), clusterName)
```
Comment (RainbowMango) [inline test/e2e/tainttoleration_test.go:72]:
```suggestion
						klog.Infof("Adding taints(%s) to cluster(%s)", taint.ToString(), clusterName)
```

File: test/e2e/tainttoleration_test.go [hunk 2]
```diff
@@ -59,26 +60,34 @@ var _ = ginkgo.Describe("propagation with taint and toleration testing", func()
...
-					err = controlPlaneClient.Update(context.TODO(), clusterObj)
-					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+					gomega.Eventually(func() bool {
+						clusterObj := &clusterv1alpha1.Cluster{}
+						err := controlPlaneClient.Get(context.TODO(), client.ObjectKey{Name: clusterName}, clusterObj)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+						clusterObj.Spec.Taints = append(clusterObj.Spec.Taints, taints...)
+						klog.Infof("update taints(%s) to clusterName(%s)", clusterObj.Spec.Taints, clusterName)
```
Comment (RainbowMango) [inline test/e2e/tainttoleration_test.go:81]:
```suggestion
						klog.Infof("add taints(%s) to cluster(%s)", clusterObj.Spec.Taints, clusterName)
```

File: test/e2e/tainttoleration_test.go [hunk 3]
```diff
@@ -92,14 +101,31 @@ var _ = ginkgo.Describe("propagation with taint and toleration testing", func()
...
+						},
+					}
+
+					gomega.Eventually(func() bool {
+						clusterObj := &clusterv1alpha1.Cluster{}
+						err := controlPlaneClient.Get(context.TODO(), client.ObjectKey{Name: clusterName}, clusterObj)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+						clusterObj.Spec.Taints = removeTargetFromSource(clusterObj.Spec.Taints, taints)
+						klog.Infof("update taints(%s) to clusterName(%s)", clusterObj.Spec.Taints, clusterName)
```
Comment (RainbowMango) [inline test/e2e/tainttoleration_test.go:119]:
```suggestion
						klog.Infof("remove taints(%s) from cluster(%s)", clusterObj.Spec.Taints, clusterName)
```

File: test/e2e/tainttoleration_test.go [hunk 4]
```diff
@@ -124,3 +147,21 @@ var _ = ginkgo.Describe("propagation with taint and toleration testing", func()
 		})
 	})
 })
+
+func removeTargetFromSource(source, target []corev1.Taint) []corev1.Taint {
+	var result []corev1.Taint
+	for si := range source {
+		flag := false
```
Comment (RainbowMango) [inline test/e2e/tainttoleration_test.go:154]:
```suggestion
		deleted := false
```

File: test/e2e/tainttoleration_test.go [hunk 5]
```diff
@@ -92,14 +99,30 @@ var _ = ginkgo.Describe("propagation with taint and toleration testing", func()
...
-					err = controlPlaneClient.Update(context.TODO(), clusterObj)
-					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+				for _, clusterName := range clusterNames {
+					taints := []corev1.Taint{
+						{
+							Key:    tolerationKey,
+							Value:  clusterName,
+							Effect: corev1.TaintEffectNoSchedule,
+						},
+					}
```
Comment (RainbowMango) [inline test/e2e/tainttoleration_test.go:109]: Can we share the `trains with the one in `BeforeEach`?

**Review by RainbowMango — CHANGES_REQUESTED** [review]

File: test/e2e/tainttoleration_test.go (hunk 4 above)
Comment (XiShanYongYe-Chang) [reply test/e2e/tainttoleration_test.go]: done

File: test/e2e/tainttoleration_test.go (hunk 5 above)
Comment (XiShanYongYe-Chang) [reply test/e2e/tainttoleration_test.go]: Here `clusterName` is a variable, I change to use 
```
func getReadyToAddTaints(tolerationKey, clusterName string) []corev1.Taint {
	return []corev1.Taint{
		{
			Key:    tolerationKey,
			Value:  clusterName,
			Effect: corev1.TaintEffectNoSchedule,
		},
	}
}
```
to get taints.

Comment (XiShanYongYe-Chang) [reply test/e2e/tainttoleration_test.go]: rename func name to `constructAddedTaints`.

**Pull request merged**